### DISCUSSION TITLE
State dependent values?

### DISCUSSION THREAD

**Discussion by harryy2510** [pmndrs/zustand#1221] [Q&A]
Hi everyone

First of all thank you creating such a great library.

I have a use-case where I need a value which is made of two values from the state. I tried two things that didn't work.

Try 1 (Doesn't works):
```
create((set, getState) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  fullName: `${getState()?.firstName} ${getState()?.lastName}`,
})
```

Try 2 (Doesn't works):
```
create((set, getState) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  get fullName() {
    return `${getState()?.firstName} ${getState()?.lastName}`
  },
})
```

The only workaround was to make `fullName` as a function. but that makes the code a little weird.

Try 3 (This works):
```
create((set, getState) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  fullName() {
    return `${getState()?.firstName} ${getState()?.lastName}`
  },
})

console.log(useStore().fullName()) // 'Hariom Sharma'
```

Is there anything else I can do about it?

**Comment by dai-shi** [comment]
What people usually do is to use a selector function.
```js
const usePersonStore = create((set) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  // ...
}))

const useFullname = () => usePersonStore((state) => `${state.firstName} ${state.lastName}`)
```

If that doesn't work, what would be a problem? There are some approaches to different problems.

**Reply by harryy2510** [reply]
Hi @dai-shi 

So for us, we wanted to avoid doing that over n over cause this property would be used quite alot.

I was thinking more of the approach where we can add virtual field based on other state values. I've seen this approach in other backend libraries like mongoose or typeorm, so just wanted to try that out if we support it. But I guess it might be a little difficult or tricky right? as we are also thinking about how to avoid re-rendering, make it performant as well?

Would like to hear your thoughts on it :)

**Reply by dai-shi** [reply]
`useFullname` is a reusable function. You don't need to do it over and over.

The selector function
```js
(state) => `${state.firstName} ${state.lastName}`
```
is very lightweight and would never be a bottleneck. If it does, you can consider memoizing it.

Now, I assume your concern is not performance, but you want some store based solutions.

### Just define a selector in a store

```js
const usePersonStore = create((set, get) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  getFullname: (state) => `${state.firstName} ${state.lastName}`,
  getFullname: () => `${get().firstName} ${get().lastName}`, // this works too
  // ...
}))

const useFullname = () => usePersonStore(usePersonStore.getState().getFullname)
```

### Define fullname property and update everything in actions

```js
const usePersonStore = create((set, get) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  fullname: 'Hariom Sharma',
  setFirstname: (firstName) => {
    set({ firstName, fullName: `${firstName} ${get().lastName}` })
  },
  // ...
}))
```

btw, this ☝️  is most performant.

### Use 3rd-party libraries

For example: https://github.com/cmlarsen/zustand-middleware-computed-state

**Reply by harryy2510** [reply]
@dai-shi Thanks again. This has been super helpful. Yeah I missed the point where I could have created a middle ware for sure.

So I do understand that this is going to be super performant.

```
const usePersonStore = create((set, get) => ({
  firstName: 'Hariom',
  lastName: 'Sharma',
  fullname: 'Hariom Sharma',
  setFirstname: (firstName) => {
    set({ firstName, fullName: `${firstName} ${get().lastName}` })
  },
  // ...
}))
```


but would the middleware be performant too? (as it does update the state in one go)
https://github.com/cmlarsen/zustand-middleware-computed-state

**Reply by dai-shi** [reply]
Yes, it should. It has only one additional function call.
